File: connection/bosh.py
from datetime import datetime, timedelta
from Queue import Queue as Pending, Empty
from threading import Timer

from notification.models import Delivery
from models import Session, Manager, Handler, getInternalRequest

import logging
import settings

logger = logging.getLogger('curve')

class Schema(object):
    """
    BOSH scheme was documented as an xml scheme in XEP-0124:
    http://www.xmpp.org/extensions/xep-0124.html
    """
    attributes = {
	'accept': str,
	'ack': int, # positive integer
	'authid': str,
	'charsets': str, # xs:NMTOKENS
	'condition': ( # xs:NCName
	    'bad-request',
	    'host-gone',
	    'host-unkonwn',
	    'improper-addressing',
	    'internal-server-error',
	    'item-not-found',
	    'other-request',
	    'policy-violation',
	    'remote-connection-failed',
	    'remote-stream-error',
	    'see-other-uri',
	    'system-shutdown',
	    'undefined-condition',
	    ),
	'content': str,
	'from': str,
	'hold': int, # unsigned byte
	'inactivity': int, # unsigned short
	'key': str,
	'maxpause': int, # unsigned short
	'newkey': str,
	'pause': int, # unsigned short
	'polling': int, # unsigned short
	'report': int, # positive integer
	'requests': int, # unsigned byte
	'rid': int, # positive integer
	'route': str,
	'sid': str,
	'stream': str,
	'time': int, # unsigned short
	'to': str,
	'type': ( # xs:NCName
	    'error',
	    'terminate',
	    ),
	'ver': str,
	'wait': int, # unsigned short
	'lang': str, # xml:lang
	}

    def check(self):
	""" Check if instance conforms to its schema. """
	if not hasattr(self, 'schema'):
	    raise TypeError('schema not found.')

	if self.schema is None:
	    raise ValueError('schema is None.')

	for name in self.schema:
	    if name not in self.attributes:
		raise ValueError('invalid schema.')
	    value = self.schema[name]
	    if value is None:
		raise ValueError('schema item:' + name + ' has no value.')
	    checkType = self.attributes[name]
	    checkType = type(checkType) == 'type' or 'string'
	    if type(value) != checkType:
		raise TypeError('schema item: ' + name + ' has invalid value type.')
	    pass
	pass
    pass

class Terminate(Schema):
    """ Server terminate session. """
    schema = [
	'type',
	'condition', # optional
	]

    def create(self, condition=None):
	""" Create server terminate response. """
	response = {
	    'type': 'terminate'
	    }
	if condition is not None:
	    response['condition'] = condition
	return response
    pass

class Connect(object):
    """
    BOSH session initial request.
    """
    schema = [
        'rid',
        'to',
        'wait',
        'ver',
        'hold',
        'lang',
        'charsets',
        ]

    def create(self, req):
        """ Create connect request. """
        request = {}
        for key in self.schema:
            if key not in req:
                logger.debug('key: ' + key + ' not found in connect request.')
                raise KeyError(key + ' not found in request.')
            else:
                request[key] = req[key]
            pass
        pass
        return request
    pass

class ConnectResponse(object):
    """
    BOSH session initial request response.
    """
    schema = [
	'sid',
	'wait',
	'ver',
	'polling',
	'inactivity',
	'requests',
	'hold',
	'to'
	]

    def create(self, sid, rid, to):
        """ Create connect response. """
	response = {
	    'sid': sid,
	    'rid': rid,
	    'to': to,
	    }
	return response
    pass

# session status.
SessionStatus = {
    'created': 0,
    'connection': 1,
    'connected': 2,
    'disconnected': 3,
    }

class BoshHandler(Handler):
    """
    Abstract BOSH session data handler.
    Classes extends this class must implementst handle interface,
    which is responsible for processing data transfered in BOSH session.
    Data passed in is an impact HTTP request.
    """
    def __init__(self, regex):
        if regex is None or type(regex) != str:
            raise ValueError('given regular expression is invalid.')
        self.regex = regex # regular expression for matching BOSH session's destionation.
        pass

    def handle(self, request):
        """
        Handle BOSH session's HTTP request, it comes from client's poll request.
        """
        raise NotImplementedError()

class BoshManager(Manager):
    """
    BOSH connection manager.
    This class implement XMPP Bidirectional-streams Over Synchronous HTTP (BOSH),
    which was defined in XEP-0124, http://xmpp.org/extensions/xep-0124.html
    """
    def connect(self, request, method='POST'):
        """
        BOSH connection manager connect implementation.

        If request contains session id, key 'sid' in request.POST, the rquest is intented
        to initizlied session, otherwise the request is a data transition.
    
        If request contains data which client sent to server, connection manager should handle it.
        And connection manager would not resposne to this request until there's data to send to
        client (recommended in BOSH).
        """
        user = request.session['userid']
        req = getInternalRequest(request, method)
        if 'sid' not in req:
            # request to initialize session
            logger.debug('request create session, userid: %s, rid: %s, to: %s', \
                             user, req['rid'], req['to'])
            # result = super(BoshManager, self).accept(user, request)
            result = self.accept(user, request, method)
        else:
            sid = req['sid']
            logger.debug('new request, sid: %s, rid: %s.', sid, req['rid'])
            try:
                session = super(BoshManager, self).getSession(sid)
                result = session.recv(request)
            except KeyError:
                result = Terminate().create('item-not-found')
                pass
            pass
        return result

    def createSession(self, sid, request, method='POST'):
	"""
	Create a bosh session.
	Connection manager SHOULD response to request.
	"""
        try:
            session = BoshSession()
            return session, session.create(sid, request, method)
        except KeyError:
            return None, Terminate().create('bad-request')
    pass

class BoshSession(Session):
    """
    BOSH session implementation.
    """
    manager = BoshManager()
    
    def __init__(self, options=settings.SESSION_OPTIONS):
        self.options = {}
        # copy options.
        for option in options:
            # logger.debug('setting options, key: %s', option)
            self.options[option] = options[option]
            pass

        self.hold = [] # holding connections.
        self.lastpoll = None # last poll timestamp
        self.lastrid = None # last rid.
        self.last = None # last data sent.
        self.inactive = None # most recent inactive timestamp.
        self.pending = Pending() # pending data queue, they will be pushed to client in a future request.
	pass

    def pollInternal(self):
        """
        Perform poll.
        If wait is not None and there's no data queued, caller will block in a duration
        of wait time (in seconds). Once poll returned, another unblocking fetching will
        be performed to check if more data queued, if so fetch rest of queued data as well.
        """
        pendings = []
        wait = self.options['wait']
        if wait is not None:
            logger.debug('try to poll on holding connect may block in %d seconds.', wait)
            try:
                data = self.pending.get(True, wait) # get queued data with block.
                pendings.append(data)
            except Empty:
                # timeout
                logger.debug('poll timeout in %d seconds.', wait)
                return self.timeout()
            pass
        
        # got head of queue, check rest of queue without blocking.
        while True:
            try:
                data = self.pending.get(False) # get queued data without block.
                pendings.append(data)
                pass
            except Empty:
                break
            finally:
                self.pending.task_done()
                pass
            pass
        # return joined queued data
        return pendings
    
    def createInternal(self, req):
	"""
	Create session and return response.
	"""
        connect = Connect().create(req)
        self.rid = connect['rid']
        self.to = connect['to']
	self.status = SessionStatus['connected']
        try:
            response = {
                'sid': self.sid,
                'rid': self.rid,
                'to': self.to,
                'ver': self.options['ver'],
                'lang': self.options['lang'],
                'wait': self.options['wait'],
                'hold': self.options['hold'],
                'inactivity': self.options['inactivity'],
                'polling': self.options['polling'],
            }
            # set create time as first poll time, to prevent client connected and hang it.
            self.lastpoll = datetime.now()
            self.scheduleInactivityCheck()
            return response
        except:
            import sys, traceback
            exc_type, exc_value, exc_traceback = sys.exc_info()
            traceback.print_exception(exc_type, exc_value, exc_traceback, \
                              limit=2, file=sys.stdout)
            logger.debug('failed to create response.')
            return None

    def scheduleInactivityCheck(self):
        """ 
        Start a timer to check if session's inactivity.
        Timer will timeout in a duration which is set in options.
        """
        Timer(self.options['inactivity'], self.checkInactivity).start()
        pass

    def checkInactivity(self):
        """
        Check session's inactivity.
        """
        now = datetime.now()
        if now > self.lastpoll + timedelta(self.options['inactivity']) and len(self.hold) < 1:
            # Connection manager SHOULD assume client disconnected.
            self.manager.disconnected(self)
        pass
    
    def send(self, data):
	"""
	Try to send data in session.
        This method simply put data in queue, if there's any connections holding,
        the head of those hold conntions will unblock.
	"""
        self.pending.put(data)
	pass

    def recv(self, request, method='POST'):
	"""
	Try poll data from connection.
        Session will trial data from request first.
        And then push queued data back to request if present, otherwise hold this connection
        until there's data ready to push to client or encountered a timeout (meaning there's 
        no data in the duration of wait time.

        This implementation will return a json serializable object.
	"""
        req = getInternalRequest(request, method)
        try:
            rid = req['rid']
        except KeyError:
            logger.debug('rid not found in request.')
            # Server will close session caused by bad request.
            self.close(condition='bad-request')
            pass

        if self.lastpoll is not None and datetime.now() < self.lastpoll + timedelta(seconds=self.options['polling']):
            # polling to frequently, overactivity.
            self.close(condition='policy-volation')
            pass

        self.lastpoll = datetime.now()
        # trial data from request
        if 'data' in req:
            self.manager.recv(req['data'])

        if rid == self.lastrid:
            # request was resent by client on broken connection.
            # server SHOULD send last HTTP 200 and queued data, see XEP-1024 broken connection.
            return self.last
        elif rid < self.lastrid:
            self.close(condition='bad-request')

        if len(self.hold) >= self.options['hold']:
            # Cant hold any connections.
            self.close(condition='policy-violation')
            pass

        # poll pending data.
        polled = self.pollInternal()
        logger.debug('poll done, got:%s.', str(polled))

        # save last rid and last data, in case of broken connection.
        self.lastrid = rid
        self.last = polled

        if len(self.hold) < 1:
            # if no connection holds, schedule an inactivity check.
            self.scheduleInactivityCheck()
            pass
        return polled
    
    def closeInternal(self, condition=None):
	""" Close BOSH session. """
	self.status = SessionStatus['disconnected']
        try:
            connection = self.hold.pop()
            terminate = Terminate().create(condition)
	    connection.send(terminate)
            # When close session, all holding connection should be closed gracefully.
            while not self.hold.empty():
                try:
                    self.hold.pop().send(terminate)
                except:
                    logger.debug('no more connection held.')
                    break
                pass
            pass
        except:
            logger.debug('no connection held.')
        pass

    def timeout(self):
        """
        When session timeouted, response an empty response to client.
        """
        return {}
    
    pass

class BoshDelivery(Delivery):
    """ Bosh delivery employee BOSH to send given data. """
    def delivery(self, data):
	""" Delivery data. """
	target = getattr(data, 'target', None)
	if target is None:
	    logger.debug('Delivery has no target, discard.')
	    return
	session = BoshSession.manager.getByOwnerId(target.user.id)
	if session is not None:
	    # target is online
	    session.send(data)
	pass
    pass

File: shared-static/js/bosh.js

!(function () {
	// "use strict";
    // BOSH scheme was documented as an xml scheme in XEP-0124:
    // http://www.xmpp.org/extensions/xep-0124.html
	var
	Connection = function () {},
	BOSHProtocol = function () {};

	// BOSH protocl.
	BOSHProtocol.prototype = {
		// This implemenation version.
		ver: '0.1',
		// Default language support.
		lang: 'en',
		// Default encoding character set.
		charset: 'UTF-8',
		// Implementation schema.
		schema: {
			accept: 'string',
			ack: 'number', // positive integer
			authid: 'string',
			charsets: 'string', // xs:NMTOKENS
			condition: [ // xs:NCName
				'bad-request',
				'host-gone',
				'host-unkonwn',
				'improper-addressing',
				'internal-server-error',
				'item-not-found',
				'other-request',
				'policy-violation',
				'remote-connection-failed',
				'remote-stream-error',
				'see-other-uri',
				'system-shutdown',
				'undefined-condition',
			],
			content: 'string',
			from: 'string',
			hold: 'number', // unsigned byte
			inactivity: 'number', // unsigned short
			key: 'string',
			maxpause: 'number', // unsigned short
			newkey: 'string',
			pause: 'number', // unsigned short
			polling: 'number', // unsigned short
			report: 'number', // positive integer
			requests: 'number', // unsigned byte
			rid: 'number', // positive integer
			route: 'string',
			sid: 'string',
			stream: 'string',
			time: 'number', // unsigned short
			to: 'string',
			type: [ // xs:NCName
				'error',
				'terminate',
			],
			ver: 'string',
			wait: 'number', // unsigned short
			lang: 'string', // xml:lang
		},

		// initial session schema.
		initialize: [
			'rid',
			'to',
			'wait',
			'ver',
			'hold',
			'lang',
			'charsets',
		],

		// request schema.
		request: [
			'sid',
			'rid',
		],

		// terminate schema.
		terminate: [
			'sid',
			'rid',
			'type',
		],

		// create BOSH request.
		create: function (schema, data) {
			return this.validate(schema, data);
		},

		// check if given data is valid.
		// This method always return data, if any errors occurred,
		// it throws an Error.
		validate: function (schema, data) {
			var index, key, type;

			if (!data) {
				throw new Error('Given data is empty.');
			}

			for (index in schema) {
				key = schema[index]
				type = this.schema[key];
				// check given schema.
				if (!type) {
					$.debug(key, ' is not a valid attribute.');
					throw new Erorr(key + ' is not a valid attribute.');
				}
				// check if data conforms to given schema.
				if (!data[key]) {
					$.debug('missing attribute:', key);
					throw new Error('Missing attribute: ' + key);
				}
				// check if data type conforms to BOSH schema.
				if (typeof(data[key]) !== type) {
					$.debug(key, ' has invalid value type.');
					throw new Error(key + ' has invalid value type.');
				}
			}
			return data;
		}
	};

	// BOSH session.
	// Instance may trigger events: "received.session", "sent.session"
	Connection.prototype = {
		// default http request method.
		method: 'post',

		// status.
		Status: {
			created: 0,
			connecting: 1,
			connected: 2,
			disconnected: 3,
		},

		// requests.
		Requests: {
			clean: 0, // no active connections.
			polling: 1, // polling from server.
		},

		// default options.
		options: {
			wait: 60, // max time to wait in seconds.
			requests: 2,
			polling: 30, // by default, client doe not perform pure polling.
		},

		// create session.
		create: function (to, options) {
			this.sid = undefined;
			this.to = to;
			this.options = $.extend({}, options);
			this.protocol = new BOSHProtocol();
			this.status = this.Status.created;
			this.handlers = [];
		},

		// generate a request id.
		nextrid: function () {
			this.rid = this.rid || (10000 + Math.round(Math.random() * 10000));
			return ++this.rid;
		},

		// Connect to host.
		connect: function (host) {
			var that = this, deferred = $.Deferred(),
			request = this.protocol.create(this.protocol.initialize, {
				to: host,
				rid: this.nextrid(),
				wait: this.options.wait,
				hold: this.options.hold,
				ver: this.options.ver,
				lang: this.options.lang,
				charsets: this.options.charsets
			});

			this.status = this.Status.connecting;

			// $[this.method](host, request, null, 'json')
			// 	.done(function (data) {
			// 		deferred.resolve(that, data);
			// 	})
			// 	.fail(function (data) {
			// 		deferred.reject(that, data);
			// 	});
			deferred.reject(that, 'disabled for debugging...');

			deferred.done(this.connectDone).fail(this.connectFailed);
			return deferred.promise();
		},

		// Add a data handler.
		// When connection received data, handler will be notified.
		addHandler: function (handler) {
			handler && this.handlers.indexOf(handler) === -1 && this.handlers.push(handler);
		},

		// If session is currently connected.
		isConnected: function () {
			return this.status === this.Status.connected;
		},

		// send data to BOSH connection manager.
		send: function (data) {
			var that = this,
			deferred = $.Deferred(),
			data = data || {},
			request = this.protocol.create(this.protocol.request, {
				sid: this.sid,
				rid: this.nextrid(),
				data: data
			});

			if (!this.isConnected()) {
				throw new Error('Connection is net connected.');
			}

			if (this.current) {
				if (this.current >= this.options.requests - 1) {
					$.debug("Can't send data right now, pending...");
					this.pending = this.pending || [];
					this.pending.push(request);
					// pending data will not be sent right now and it will not trigger 'sent.session' event.
					return deferred.reject('pending');
				} else {
					// increase current requests.
					++this.current;
				}
			} else {
				this.current = this.Requests.polling;
			}

			request = request || {};
			$[this.method](this.to, request)
				.done(function (response) {
					deferred.resolve(that, response);
				})
				.fail(function (response) {
					deferred.reject(that, response);
				})
				.always(function () {
					// that.trigger('sent.session');
				});

			deferred.done(this.onReceive).fail(this.onBroken);
			return deferred.promise();
		},

		// Handle http connection response (receive data from server in BOSH).
		onReceive: function (connection, response) {
			var i, request;

			response = typeof response === 'string' ? JSON.parse(response) : response;
			if (response && !$.isEmptyObject(response)) {
				// Potensial problem, if connection is receiving data,
				// and a new handler added, that handler may results differently,
				// it may or may not receive data in this transition.
				for (i = 0; i < connection.handlers.length; i++) {
					connection.handlers[i].call(connection, response);
				}
			}

			// decrease current requests.
			--connection.current;

			// client SHOULD send another request to poll incoming data immediately.
			switch (connection.current) {
			case connection.Requests.clean:
				// if session is currently clean, send another request,
				// the request to be sent will be empty if no data pending,
				// or the head data in pending list.
				request = connection.pending && connection.pending.pop();
				$.debug(request ? 'sending request...' : 'keep polling...');
				connection.send(request || {});
				break;
			case connection.Requests.polling:
			default:
				// if connection is currently polling. do nothing.
				break;
			}
		},

		// Handle send failed due to broken connection.
		onBroken: function (connection, response) {
			connection.close();
		},

		// Receive data from connection.
		// Since data always received in send response (see BOSH),
		// this method only trial data already received in send response.
		recv: function () {
			if (!this.isConnected()) {
				throw new Error('connection is not connected.');
			}
			return this.received && this.received.pop()
		},

		// Close connection, sent terminate request to server.
		// furhter data received from server will be discarded.
		close: function () {
			var terminate = {
				sid: this.sid,
				rid: this.nextrid(),
				type: 'terminate'
			};
			return $[this.method](this.to, terminate)
				.done()
				.fail()
				.always(this.cleanup);
		},

		// clean up connection after it was closed.
		cleanup: function () {
			if (this.received) {
				while (this.received.length) {
					this.received.pop();
				}
				delete this.received;
			}
		},

		// Handle connection connect done.
		connectDone: function (connection, response) {
			var connection, to = response.to;

			if (response.type && response.type === 'terminate') {
				// connection failed to create due to server terminated.
				$.debug('connection failed to connect.');
				connection.status = connection.Status.disconnected;
				connection.cleanup();
				return;
			}

			$.debug('Connected to:', to);
			connection.status = connection.Status.connected;
			connection.sid = response.sid;
			$.extend(connection.options, {
				wait: Math.min(response.wait, connection.options.wait),
				requests: Math.max(response.requests, connection.options.requests),
				polling: response.polling || connection.options.polling,
			});

			// since connection established, start polling.
			connection.send();
		},

		// Handle session connect failed.
		// Manager may want to reconnect in a future.
		connectFailed: function (connection, response) {
			$.debug('Faield to connect to:', response ? response.to : 'no response');
			connection.status = connection.Status.disconnected;
			connection.cleanup();
		}
	};

	// register local manager as curve's component.
	$.curve.component('CM', {
		// default options for all session.
		options: {
			ver: '0.1',
			hold: 1,
			lang: 'en',
			charsets: 'UTF-8',
			wait: 60,
			requests: 2,
			polling: 30,
		},

		// create component.
		create: function () {
			this.connections = [];
		},

		// Get connection.
		get: function (host) {
			if (host in this.connections) {
				return this.connections[host];
			} else {
				$.debug('connection matches:', host, ' not found.');
			}
		},

		// Try to connect to server to create a connection.
		connect: function (host) {
			var connection = new Connection();
			connection.create(host, this.options);
			// save connection, notice that sesion not fully created yet.
			this.connections[host] = connection;
			return connection.connect(host);
		},

		// Try to close current connection.
		disconnect: function (host) {
			var connection = this.connections[host];
			if (!connection) {
				$.debug('Connection matches:', host, ' not found');
				throw new Error('connection not found.');
			} else {
				if (connection.isConnected()) {
					connection.close().always(function () {
						delete this.connections[host];
					});
				} else {
					delete this.connections[host];
				}
			}
		}
	});
}($));
